src/shogun/distributions/classical/GaussianDistributionNew.cpp

```diff
@@ -0,0 +1,297 @@
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ * The views and conclusions contained in the software and documentation are those
+ * of the authors and should not be interpreted as representing official policies,
+ * either expressed or implied, of the Shogun Development Team.
+ *
+ */
+
+
+// Do we need to initialise m_dimension or is it somehow done automatically?
+// How should we store (i.e. under which type) m_cov_type?
+
+#include <shogun/distributions/classical/GaussianDistributionNew.h>
+
+#ifdef HAVE_EIGEN3
+
+#include <shogun/base/Parameter.h>
+#include <shogun/mathematics/eigen3.h>
+
+using namespace shogun;
+using namespace Eigen;
+
+CGaussianDistributionNew::CGaussianDistributionNew() : CProbabilityDistribution()
+{
+	init();
+}
+
+CGaussianDistributionNew::CGaussianDistributionNew(SGVector<float64_t> mean,
+		float64_t cov, bool cov_is_cholesky) :
+				CProbabilityDistribution(mean.vlen)
+{
+	init();
+
+	m_mean=mean;
+
+	m_cov=SGMatrix<float64_t>(1,1);
+	m_cov(0,0)=cov;
+
+	m_cov_type="Float";
+}
+
+CGaussianDistributionNew::CGaussianDistributionNew(SGVector<float64_t> mean,
+		SGVector<float64_t> cov, bool cov_is_cholesky) :
+				CProbabilityDistribution(mean.vlen)
+{
+	REQUIRE(mean.vlen==cov.vlen,
+			"Mean dimension (%d) and covariance dimension (%d) do not match\n",
+			mean.vlen, cov.vlen);
+
+	init();
+
+	m_mean=mean;
+
+	m_cov=cov.convert_to_matrix(cov,1,cov.vlen,false);
+
+	m_cov_type="Vector";
+}
+
+CGaussianDistributionNew::CGaussianDistributionNew(SGVector<float64_t> mean,
+		SGMatrix<float64_t> cov, bool cov_is_cholesky) :
+				CProbabilityDistribution(mean.vlen)
+{
+	REQUIRE(cov.num_rows==cov.num_cols, "Covariance must be square but is "
+			"%dx%d\n", cov.num_rows, cov.num_cols);
+	REQUIRE(mean.vlen==cov.num_cols,
+			"Mean dimension (%d) and covariance dimension (%d) do not match\n",
+			mean.vlen, cov.num_cols);
+
+	init();
+
+	m_mean=mean;
+
+	if (!cov_is_cholesky)
+	{
+		compute_cholesky(cov);
+	}
+	else
+		m_cov=cov;
+
+	m_cov_type="Matrix";
+}
+
+CGaussianDistributionNew::~CGaussianDistributionNew()
+{
+
+}
+
+void CGaussianDistributionNew::set_mean(SGVector<float64_t> mean)
+{
+	m_mean=mean;
+}
+
+void CGaussianDistributionNew::set_cov(float64_t cov)
+{
+	m_cov=SGMatrix<float64_t>(1,1);
+	m_cov(0,0)=cov;
+
+	m_cov_type="Float";
+}
+
+void CGaussianDistributionNew::set_cov(SGVector<float64_t> cov)
+{
+	m_cov=cov.convert_to_matrix(cov,1,cov.vlen,false);
+
+	m_cov_type="Vector";
+}
+
+void CGaussianDistributionNew::set_cov(SGMatrix<float64_t> cov)
+{
+	m_cov=cov;
+
+	m_cov_type="Matrix";
+}
+
+SGVector<float64_t> CGaussianDistributionNew::get_mean() const
+{
+	return m_mean;
+}
+
+float64_t CGaussianDistributionNew::get_cov_spherical() const
+{
+	return *m_cov.data();
+}
+
+SGVector<float64_t> CGaussianDistributionNew::get_cov_diag() const
+{
+	return m_cov.get_row_vector(0);
```

Comment:
I think this copies the vector, 
http://shogun-toolbox.org/doc/en/latest/SGMatrix_8cpp_source.html#l01088
but it would be better to return the same memory (this should be documented in the getter doxygen)

Shogun stores vectors in column-major (fortran) format, which means you can extract subsequent columns of matrices without copying. So get_column_vector is your friend here:
http://shogun-toolbox.org/doc/en/latest/SGMatrix_8h_source.html#l00115
Just realised there is no version of it which returns SGVector, so you could add that for convenience. Ask me if unclear

Summary:
We dont want to copy here, but return a reference to the memory. And this should be stated in the documentation